frigate/embeddings/maintainer.py

```diff
@@ -0,0 +1,180 @@
+                camera_config = self.config.cameras[camera]
+                if data["id"] not in self.tracked_events:
+                    self.tracked_events[data["id"]] = []
+
+                # Create our own thumbnail based on the bounding box and the frame time
+                try:
+                    frame_id = f"{camera}{data['frame_time']}"
+                    yuv_frame = self.frame_manager.get(
+                        frame_id, camera_config.frame_shape_yuv
+                    )
+                    data["thumbnail"] = self._create_thumbnail(yuv_frame, data["box"])
+                    self.tracked_events[data["id"]].append(data)
+                    self.frame_manager.close(frame_id)
+                except FileNotFoundError:
+                    continue
+
+            # Embed thumbnails when an event ends
+            while True:
+                ended = self.event_end_subscriber.check_for_update()
+
+                if ended == None:
+                    break
+
+                event_id, camera, updated_db = ended
+                camera_config = self.config.cameras[camera]
+
+                if updated_db:
+                    try:
+                        event: Event = Event.get(Event.id == event_id)
+                    except DoesNotExist:
+                        continue
+
+                    # Skip the event if not an object
+                    if event.data.get("type") != "object":
+                        continue
+
+                    # Extract valid event metadata
+                    metadata = get_metadata(event)
+                    thumbnail = base64.b64decode(event.thumbnail)
+
+                    # Embed the thumbnail
+                    self._embed_thumbnail(event_id, thumbnail, metadata)
+
+                    if (
+                        camera_config.genai.enabled
+                        and self.genai_client is not None
+                        and event.data.get("description") is None
+                    ):
+                        # Generate the description. Call happens in a thread since it is network bound.
+                        threading.Thread(
+                            target=self._embed_description,
+                            name=f"_embed_description_{event.id}",
+                            daemon=True,
+                            args=(
+                                event,
+                                [
+                                    data["thumbnail"]
+                                    for data in self.tracked_events.get(
+                                        event_id, [{"thumbnail": thumbnail}]
+                                    )
+                                ],
+                                metadata,
+                            ),
+                        ).start()
+
+                # Delete tracked events based on the event_id
+                if event_id in self.tracked_events:
+                    del self.tracked_events[event_id]
```

Comment:
I'll break the logic out into additional methods which should make this a lot easier to follow.

I'm purposefully using two subscribers here for the reason stated above:
> The update publisher publishes all event updates based on object processing. This includes start/update/end.
The event end publisher is based on the event maintainer and gets published only after the finished event is saved to the DB (or not).

There shouldn't be a race condition since we process all updates first which include start/update/end, and the end subscriber is only after it's saved to the DB. I need the DB record as a fallback to get the thumbnail for things like manual events where we aren't tracking the detection frames for "significant updates".